File: 0377-combination-sum-iv/0377-combination-sum-iv.py
class Solution:
    def combinationSum4(self, nums: List[int], target: int) -> int:
        dp = [0] * (target + 1)
        dp[target] = 1
        for i in range(target-1, 0, -1):
            for num in nums:
                if i + num <= target:
                    dp[i] += dp[i + num]
        ans = 0
        for num in nums:
            if num <= target:
                ans += dp[num]
        return ans

File: 0377-combination-sum-iv/0377-combination-sum-iv.cpp
class Solution {
public:
    int combinationSum4(vector<int>& nums, int target) {

        vector<unsigned int> dp(target + 1, 0);
        dp[target] = 1;
        for (int i = target - 1; i > 0; i--) {

            for (int num : nums) {
                if (num + i <= target) { dp[i] += dp[i + num]; }

            }

        }
        int ans = 0;
        for (int num : nums) {
            if (num <= target) { ans += dp[num]; }
        }

        return ans;
    }
};